### ISSUE TITLE
SSH still available in container even I disable `forwardWSLServices`

### ISSUE THREAD

**Issue by ai** [microsoft/vscode-remote-release#9897]
SSH inside container can connect to GitHub even if I disable `forwardWSLServices` and `gitCredentialHelperConfigLocation`.

There are no keys in `~/.ssh` (inside container) and no `helper` in `git config -l`

My settings:

```
"dev.containers.copyGitConfig": false,
  "dev.containers.defaultExtensions": [
    "editorconfig.editorconfig",
    "connor4312.nodejs-testing",
    "dbaeumer.vscode-eslint",
    "esbenp.prettier-vscode",
    "yoavbls.pretty-ts-errors",
    "streetsidesoftware.code-spell-checker",
    "editorconfig.editorconfig",
    "yzhang.markdown-all-in-one",
    "christian-kohler.path-intellisense",
    "christian-kohler.npm-intellisense",
    "VisualStudioExptTeam.vscodeintellicode",
    "davidlday.languagetool-linter"
  ],
  "dev.containers.dockerCredentialHelper": false,
  "dev.containers.dockerPath": "podman",
  "dev.containers.forwardWSLServices": false,
  "dev.containers.gitCredentialHelperConfigLocation": "none",
  "dev.containers.mountWaylandSocket": false,
  "dotfiles.installCommand": "devcontainer/install-dotfiles",
  "dotfiles.repository": "ai/environment",
```

- VSCode Version: 1.89.1
- Local OS Version: Fedora 40
- Remote OS Version: ubuntu:24.04
- Remote Extension/Connection Type: Containers

- Logs:

Steps to Reproduce:

1. Disable `forwardWSLServices`, and `gitCredentialHelperConfigLocation`
2. Open local [project](https://github.com/hplush/slowreader/blob/main/.devcontainer/podman/devcontainer.json) in container
3. Open container terminal in VS Code
4.  Run `ssh -T git@github.com`
5. See container has SSH access to GitHub

**Comment by chrmarti** [comment]
`"dev.containers.forwardWSLServices": false` is only for WSL. If you have an `ssh-agent` running locally, that will still be forwarded. You could clear the `SSH_AUTH_SOCK` env variable when starting VS Code (mainly applies if your local machine is Linux).

**Comment by ai** [comment]
> The extension also forwards the X11 display, the Wayland display, the gpg-agent (if GPG is available in the container) and the Docker credential helper (if Docker is available in the container).

Yes. I try to use Dev Container to prevent been hacked from `node_module`. So I need an option to remove GPG. X11/Wayland, Docker sync.

Seems like we have UI issue. It was unclear to me what WSL is in the context of Dev Container plugin (I am using Linux). Maybe we can rename option to explain that they do not protect Linux/Mac machine.

Also, can we add options to disable forwarding `ssh-agent` (I may need `SSH_AUTH_SOCK` on local machine for another tools), `gpg-agent`, X11/Wayland? Is it possible to disable them (for experienced users of course, disabling by default will reduce DX)?

**Comment by chrmarti** [comment]
My understanding is that Docker containers are not a security boundary. So even with all these features disabled, you won't have a "secure" setup. There might be ways to harden the container I'm not familiar with.

**Comment by ai** [comment]
@chrmarti can we at least add option to disable it in JSON settings (to keep UI clean)?

I agree that it is not silver bullet, but it is at least one of isolation techniques from supply chain attack.

Supply chain risks are big problem for developers. Any developer can be hacked from `node_modules`. I want to create a multiple layers to protect myself from `node_modules`. I understand that there is no silver bullet, so I want to at least increase the price of attack.

I am using Podman (root-less Docker). Theoretically there are ways to get out from Podman container, but it means that attacker should know it and have right tools.

Right now all known attacks from supply chain (there were few dependencies stealing crypto wallets) was not ready for Docker/Podman.

This is why I think we should add some option to continue experimenting in this area. It is a very trendy topic, many companies are looking for some solution and having at least some supply chain protection could be great advantage for VS Code. If not, few extra `if` and option for JSON settings will be a very tiny things.

**Comment by ai** [comment]
BTW, Red Hat is talking about Podman as a tool for security boundary:

> A greater level of application security is possible because Podman does not run a daemon or system service requiring administrative (or sudo) rights to build and run those resources. This improves the protection of not only the applications but also the host environment in the event that a container image contains a weakness from vulnerable software packages.

— [Enhancing application container security and compliance with Podman](https://www.redhat.com/en/blog/enhancing-application-container-security-and-compliance-podman)

Podman's default setup includes rootless containers running in isolated user namespaces, limiting the impact of any potential breakout.

**Comment by tilosp** [comment]
> `"dev.containers.forwardWSLServices": false` is only for WSL. If you have an `ssh-agent` running locally, that will still be forwarded. You could clear the `SSH_AUTH_SOCK` env variable when starting VS Code (mainly applies if your local machine is Linux).

@chrmarti 
clearing the SSH_AUTH_SOCK env variable is not always an option because it can also break SSH Remotes.
